### PR TITLE
[Merged by Bors] - feat(Bicategory/Functor/Lax): define lax functors

### PR DESCRIPTION
**Pull request by callesonne**
We define lax functors and provide ways to construct pseudofunctors from a given a lax functor.

---

- [x] depends on: #14098

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+-- the simpNF linter complains that `map₂_leftUnitor_assoc` etc can be
+-- proved with `simp` so I move them here
+attribute [reassoc] map₂_leftUnitor map₂_rightUnitor
+attribute [simp] map₂_leftUnitor map₂_rightUnitor
+
+/-- The underlying prelax functor. -/
+add_decl_doc LaxFunctor.toPrelaxFunctor
+
+attribute [nolint docBlame] CategoryTheory.LaxFunctor.mapId
```
Comment (yuma-mizuno) [inline Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean:98]: At this stage it's better to have docs, at least for non-Prop fields.

Comment (callesonne) [reply Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean]: Fixed! Let me know if the description sounds reasonable, I am not an expert in bicategories.

File: Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+initialize_simps_projections LaxFunctor (+toPrelaxFunctor, -obj, -map, -map₂)
+
+namespace LaxFunctor
+
+attribute [reassoc (attr := simp)]
+  mapComp_naturality_left mapComp_naturality_right map₂_associator
+
+-- the simpNF linter complains that `map₂_leftUnitor_assoc` etc can be
+-- proved with `simp` so I move them here
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean:95]: This is a perfectly good reaction from the `simpNF` linter. The main reason is that there is no composition in the LHS.

File: Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+attribute [reassoc] map₂_leftUnitor map₂_rightUnitor
+attribute [simp] map₂_leftUnitor map₂_rightUnitor
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean:97]:
```suggestion
attribute [simp, reassoc] map₂_leftUnitor map₂_rightUnitor
```

File: Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+/-- The underlying prelax functor. -/
+add_decl_doc LaxFunctor.toPrelaxFunctor
+
+variable (F : LaxFunctor B C)
+
+@[reassoc]
+lemma mapComp_assoc_left {a b c d : B} (f : a ⟶ b) (g : b ⟶ c) (h : c ⟶ d) :
+    F.mapComp f g ▷ F.map h ≫ F.mapComp (f ≫ g) h = (α_ (F.map f) (F.map g) (F.map h)).hom ≫
+    F.map f ◁ F.mapComp g h ≫ F.mapComp f (g ≫ h) ≫ F.map₂ (α_ f g h).inv := by
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean:107]:
```suggestion
      F.map f ◁ F.mapComp g h ≫ F.mapComp f (g ≫ h) ≫ F.map₂ (α_ f g h).inv := by
```

File: Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+  rw [← F.map₂_associator_assoc, ← F.map₂_comp]
+  simp only [Iso.hom_inv_id, PrelaxFunctor.map₂_id, comp_id]
+
+@[reassoc]
+lemma mapComp_assoc_right {a b c d : B} (f : a ⟶ b) (g : b ⟶ c) (h : c ⟶ d) :
+    F.map f ◁ F.mapComp g h ≫ F.mapComp f (g ≫ h) =
+    (α_ (F.map f) (F.map g) (F.map h)).inv ≫ F.mapComp f g ▷ F.map h ≫
+    F.mapComp (f ≫ g) h ≫ F.map₂ (α_ f g h).hom := by
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean:115]:
```suggestion
      (α_ (F.map f) (F.map g) (F.map h)).inv ≫ F.mapComp f g ▷ F.map h ≫
        F.mapComp (f ≫ g) h ≫ F.map₂ (α_ f g h).hom := by
```

File: Mathlib/CategoryTheory/Bicategory/Functor/Pseudofunctor.lean [hunk 6]
```diff
@@ -260,6 +240,32 @@ noncomputable def mkOfOplax' (F : OplaxFunctor B C) [∀ a, IsIso (F.mapId a)]
     simp only [assoc, F.map₂_associator]
 #align category_theory.pseudofunctor.mk_of_oplax' CategoryTheory.Pseudofunctor.mkOfOplax'
 
+/-- Construct a pseudofunctor from a lax functor whose `mapId` and `mapComp` are isomorphisms. -/
+@[simps]
+def mkOfLax (F : LaxFunctor B C) (F' : F.PseudoCore): Pseudofunctor B C where
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Functor/Pseudofunctor.lean:245]:
```suggestion
def mkOfLax (F : LaxFunctor B C) (F' : F.PseudoCore) : Pseudofunctor B C where
```

File: Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean (hunk 2 above)
Comment (callesonne) [reply Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean:95]: Okay, I will remove it from `Oplax.lean` then as well!

File: Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean (hunk 3 above)
Comment (callesonne) [reply Mathlib/CategoryTheory/Bicategory/Functor/Lax.lean]: Made corresponding change to `Oplax.lean` and `Pseudofunctor.lean`

**Pull request closed without merging**